core/test/qa_Messages.cpp

```diff
@@ -0,0 +1,302 @@
+#include <boost/ut.hpp>
+
+#include <gnuradio-4.0/Scheduler.hpp>
+#include <magic_enum.hpp>
+#include <magic_enum_utility.hpp>
+
+#if defined(__clang__) && __clang_major__ >= 16
+// clang 16 does not like ut's default reporter_junit due to some issues with stream buffers and output redirection
+template<>
+auto boost::ut::cfg<boost::ut::override> = boost::ut::runner<boost::ut::reporter<>>{};
+#endif
+
+using namespace std::chrono_literals;
+using namespace std::string_literals;
+
+template<typename T>
+struct SimpleSource : gr::Block<SimpleSource<T>> {
+    T              value;
+    gr::PortOut<T> out;
+
+    T
+    processOne() {
+        return value;
+    }
+};
+
+ENABLE_REFLECTION_FOR_TEMPLATE(SimpleSource, out);
+
+template<typename T>
+struct CoutSink : gr::Block<CoutSink<T>> {
+    gr::PortIn<T> in;
+
+    void
+    processOne(T value) {
+        std::cout << value << "  ";
+    }
+};
+
+ENABLE_REFLECTION_FOR_TEMPLATE(CoutSink, in);
+
+// TODO:
+//  - needed: sources, basic process, sinks
+//            message senders, message checkers/receivers
+//  - test: one block sends another a message in the same graph
+//  - test: block from outsie sends a message to a block inside the graph
+//  - test: block from inside of a graph sends a message to outside
+//  - test: one block sends a multicast message to everyone in a graph
+//  - test: control messages to start and stop
+//  - test: settings update messages
+//
+//  - tests: custom message ports
+//  - processMessages overloads with 'different' message port types
+//
+//  - test: message processing when scheduler is stopped
+//
+//  idea: tests can use messages to trigger checks
+
+struct EventChecker {
+    struct Event {
+        std::string who;
+        std::string what;
+    };
+
+    std::vector<Event> _recordedEvents;
+    std::mutex         _recordedEventsMutex;
+
+    void
+    addEvent(const std::string &who, const std::string &what) {
+        fmt::print(">> Event added to the record {} -> {}\n", who, what);
+        _recordedEvents.emplace_back(who, what);
+    }
+
+    bool
+    checkEvents(std::string who, const std::vector<std::string> &whats) {
+        return std::ranges::equal(_recordedEvents                                                                       //
+                                          | std::views::filter([&who](const Event &event) { return who == event.who; }) //
+                                          | std::views::transform(&Event::what),
+                                  whats);
+    }
+};
+
+///
+
+template<typename T>
+struct CountSource : public gr::Block<CountSource<T>> {
+    gr::PortOut<T> out;
+    std::size_t    n_samples_max = 0;
+    std::size_t    count         = 0;
+
+    EventChecker *_eventChecker;
+
+    CountSource(EventChecker &eventChecker) : _eventChecker(std::addressof(eventChecker)) {}
```

Comment:
Not ideal, since the CRTP constructor is not being called. We should refrain from declaring custom constructors/destructors in user-defined sources.

Reply:
It calls Block(), which calls Block({}), but I can move it from the constructor..